Mathlib/Data/Set/Pointwise/Basic.lean

```diff
@@ -0,0 +1,1171 @@
+import Mathlib.Data.Set.Lattice
+import Mathlib.Data.Nat.Order.Basic
+
+/-!
+# Pointwise operations of sets
+
+This file defines pointwise algebraic operations on sets.
+
+## Main declarations
+
+For sets `s` and `t` and scalar `a`:
+* `s * t`: Multiplication, set of all `x * y` where `x ∈ s` and `y ∈ t`.
+* `s + t`: Addition, set of all `x + y` where `x ∈ s` and `y ∈ t`.
+* `s⁻¹`: Inversion, set of all `x⁻¹` where `x ∈ s`.
+* `-s`: Negation, set of all `-x` where `x ∈ s`.
+* `s / t`: Division, set of all `x / y` where `x ∈ s` and `y ∈ t`.
+* `s - t`: Subtraction, set of all `x - y` where `x ∈ s` and `y ∈ t`.
+
+For `α` a semigroup/monoid, `set α` is a semigroup/monoid.
+As an unfortunate side effect, this means that `n • s`, where `n : ℕ`, is ambiguous between
+pointwise scaling and repeated pointwise addition; the former has `(2 : ℕ) • {1, 2} = {2, 4}`, while
+the latter has `(2 : ℕ) • {1, 2} = {2, 3, 4}`. See note [pointwise nat action].
+
+Appropriate definitions and results are also transported to the additive theory via `to_additive`.
+
+## Implementation notes
+
+* The following expressions are considered in simp-normal form in a group:
+  `(λ h, h * g) ⁻¹' s`, `(λ h, g * h) ⁻¹' s`, `(λ h, h * g⁻¹) ⁻¹' s`, `(λ h, g⁻¹ * h) ⁻¹' s`,
+  `s * t`, `s⁻¹`, `(1 : set _)` (and similarly for additive variants).
+  Expressions equal to one of these will be simplified.
+* We put all instances in the locale `pointwise`, so that these instances are not available by
+  default. Note that we do not mark them as reducible (as argued by note [reducible non-instances])
+  since we expect the locale to be open whenever the instances are actually used (and making the
+  instances reducible changes the behavior of `simp`.
+
+## Tags
+
+set multiplication, set addition, pointwise addition, pointwise multiplication,
+pointwise subtraction
+-/
+
+
+library_note "pointwise nat action"/--
+Pointwise monoids (`set`, `finset`, `filter`) have derived pointwise actions of the form
+`has_smul α β → has_smul α (set β)`. When `α` is `ℕ` or `ℤ`, this action conflicts with the
+nat or int action coming from `set β` being a `monoid` or `div_inv_monoid`. For example,
+`2 • {a, b}` can both be `{2 • a, 2 • b}` (pointwise action, pointwise repeated addition,
+`set.has_smul_set`) and `{a + a, a + b, b + a, b + b}` (nat or int action, repeated pointwise
+addition, `set.has_nsmul`).
+
+Because the pointwise action can easily be spelled out in such cases, we give higher priority to the
+nat and int actions.
+-/
+
+
+open Function
+
+variable {F α β γ : Type _}
+
+namespace Set
+
+/-! ### `0`/`1` as sets -/
+
+
+section One
+
+variable [One α] {s : Set α} {a : α}
+
+/-- The set `1 : set α` is defined as `{1}` in locale `pointwise`. -/
+@[to_additive "The set `0 : set α` is defined as `{0}` in locale `pointwise`."]
+protected noncomputable def hasOne : One (Set α) :=
+  ⟨{1}⟩
+#align set.has_one Set.hasOne
+
+scoped[Pointwise] attribute [instance] Set.hasOne Set.hasZero
+
```

Comment:
This PR is currently stuck on this line; scoped instance attributes seem to be currently unsupported.

Reply:
To report here the upshot of this thread: this is apparently actually a bug in lean core.  PR https://github.com/leanprover/lean4/pull/2008 should fix it.

Reply:
That PR has been merged.